### DISCUSSION TITLE
[C# Markup] How should I use ThemeResource in template control?

### DISCUSSION THREAD

**Discussion by kucint** [unoplatform/uno.extensions#2268] [Q&A] [answered]
The uno documentation [Resource Dictionaries: Creating Resources](https://platform.uno/docs/articles/external/uno.extensions/doc/Reference/Markup/Resources.html#creating-resources) claims, that in order to create Theme Resource Dictionary one should write following code:

```cs
public static class MyResources
{
    public static readonly ThemeResource<Color> MyColor =
        ThemeResource.Create<Color>(nameof(MyColor), "#FF0000", "#FFFFFF");
}
```

this code produces a compilation error: 
_error CS0122: 'ThemeResource<T>' is inaccessible due to its protection level_

The `ThemeResource` class is indeed `internal sealed class`.

**Q1:** How should I create my theme resources correctly?

I suppose you have a bug in documentaion and you intended to write the following:

```cs
public static class MyResources
{
    public static readonly Resource<Color> MyColor =
        ThemeResource.Create<Color>(nameof(MyColor), "#FF0000", "#FFFFFF");
}
```

**Q2:** How should I bind themes to template control?

```cs
internal class MyControl: Control
{
    public MyControl() => this
        .DataContext(new BindableMyControlModel(), (page, vm) => page
            .Resources(r => r.Add(MyResources.MyColor))
            .Style(new MyStyle())
        );
}

class MyStyle
{
    public MyStyle() => Setters(s => s.Template(c =>
        new Border()
        .VerticalAlignment(VerticalAlignment.Top)
        .HorizontalAlignment(HorizontalAlignment.Center)
        .Width(100)
        .Height(100)
        .BorderThickness(1)
        .BorderBrush(MyResources.MyColor); //  this is how documentation suggest to make an implicit binding,
    ));
}
```

Following does not work:
- `.BorderBrush(MyResources.MyColor); ` this one does not compile
- `.BorderBrush(() => MyResources.MyColor); ` this one does compile, and runs, but has no effect
- `.BorderBrush(nameof(MyResources.MyColor)); ` this one does compile, but crashes with _System.InvalidOperationException: 'The color MyColor is unknown'_
- `.BorderBrush(x => x.ThemeResource(nameof(MyResources.MyColor))); ` this one does compile, and runs, but has no effect
- `.BorderBrush(x => x.StaticResource(nameof(MyResources.MyColor))); ` this one does compile, and runs, but has no effect

**Comment by kucint** [comment]
ok. I found the error:
I have to attach resources to the control before attaching vm (but why?)
```cs
internal class MyControl: Control
{
    public MyControl() => this
        .Resources(r => r.Add(MyResources.MyColor))                   // resources first
        .DataContext(new BindableMyControlModel(), (page, vm) => page // than vm...
            .Style(new MyStyle())
        );
}
```

... and use explicit `ThemeResource` binding: `.BorderBrush(x => x.ThemeResource(nameof(MyResources.MyColor)));`
All others ways to bind do not work....

This is a pity, because the coolest syntax that is mentioned in the documentation, that automatically generates implicit binding unfortunately does not compile:
`.BorderBrush(MyResources.MyColor));`  // does not compile for templated control

Is these two bugs? will they be fixed?

**Comment by dansiegel** [comment, marked as answer]
1) > The ThemeResource class is indeed internal sealed class

    So `ThemeResource` is a static public class. `ThemeResource<T>` which is the one you're trying to use in your sample is an internal sealed class and it's meant to be because it's the BindingProvider that is used internally by the `IDependencyPropertyBuilder` if you did `x => x.ThemeResource("Foo")`

    You are correct there is a bug in the docs. I will have a PR up this morning to fix it. Thanks for bringing it to my attention. (see #2269)
    
2) > How should I bind themes to template control?

    This one gets a little trickier. There are some edge cases for Resource resolution that are very hard for us to properly replicate from C# Markup because we don't really know the full Visual Tree when we are creating the element. We have some helpers that attempt to wait for the control to be loaded to resolve the resource, but this doesn't always work. I'm creating an issue for us to track internally to revisit this and see if we can improve the resource resolution for this scenario. Resources that are available globally in the Application will always be the easiest for the Theme/Static resources to resolve. 

3) `BorderBrush(MyResources.MyColor)` This not working would be expected because you need to think of the type here. Remember that the Resource for `MyColor` is a type `Color` while the `BorderBrush` like the names suggests is a `Brush`. While we do have some helpers here that allow you to do `BorderBrush(Colors.Red)` or `BorderBrush("#F1F1F1")` you cannot use the strongly typed Resource here by design. What you would want to look at doing would be something more like:
    ```cs
    .BorderBrush(new SolidColorBrush().Color(MyResources.MyColor))
    ```

**Reply by kucint** [reply]
@dansiegel, thanks for explanation. it is more clear to me how binding in template works.

FYI: Following seems not to work...
```cs
.BorderBrush(new SolidColorBrush().Color(MyResources.MyColor))
```
but this one indeed does!
```cs
.BorderBrush(MyResources.MyBrush))
```

**Reply by dansiegel** [reply]
Thanks for letting me know. I'll create an issue for this.
